### PR TITLE
feat: COUNT Queries

### PR DESCRIPTION
**Pull request by tom-andersen**
Add support for "count" queries; that is, a query that gets the number of documents in the result set without actually downloading the documents.

### CODE REVIEW

File: dev/src/reference.ts
Comment (dconeybe) [inline dev/src/reference.ts:2941]: Please improve this error message.

File: dev/src/reference.ts
Comment (dconeybe) [inline dev/src/reference.ts:2975]: Do you need to put a space after the ':' character? Also on line 2983 below.

File: dev/src/reference.ts
Comment (dconeybe) [inline dev/src/reference.ts:3096]: Remove getCount(); the users should just get if from data().

**Review by dconeybe — CHANGES_REQUESTED** [review]
I've reviewed the non-test code so far. I'm not done my review, but wanted to send my feedback now because I probably won't get back to it until later today.

File: dev/src/reference.ts
Comment (tom-andersen) [reply dev/src/reference.ts:2975]: This is copied code from the Query class _stream method.

I investigated, and under the covers it uses the `util.format` function that formats correctly.

File: dev/src/reference.ts
Comment (tom-andersen) [reply dev/src/reference.ts:3096]: Removed

File: types/firestore.d.ts
Comment (tom-andersen) [reply types/firestore.d.ts:1722]: Removed

File: types/firestore.d.ts
Comment (tom-andersen) [reply types/firestore.d.ts:2095]: Removed

File: dev/src/reference.ts
Comment (dconeybe) [inline dev/src/reference.ts:1618]: Explicitly change the visibility of ` _aggregate` to `private`, using the `private` keyword.

**Review by dconeybe — DISMISSED** [review]
Thank you for addressing my comments, Tom. I just have one minor follow-up comment. I'm going to pass off this review to @ehsannas or @wu-hui to also review the test code.

**Comment by ehsannas** [comment]
Can you fix the CI test failures? (`Property 'getCount' does not exist on type 'AggregateQuerySnapshot<{ count: AggregateField<number>; }>'`)

File: dev/src/reference.ts
Comment (ehsannas) [inline dev/src/reference.ts:3022]: `or read time` ?

File: dev/system-test/firestore.ts
Comment (ehsannas) [inline dev/system-test/firestore.ts:2216]: A few more tests could be added (e.g. with `startAt` and `endAt`, with `offset`, with collection groups, and in the context of transactions). Java SDK [examples](https://github.com/googleapis/java-firestore/blob/ef39a25263383303a1255a6447048e02d15da0d8/google-cloud-firestore/src/test/java/com/google/cloud/firestore/ITQueryCountTest.java).

File: dev/src/reference.ts
Comment (dconeybe) [reply dev/src/reference.ts:2865]: (drive-by comment) I *think* there is still a precedent for doing "unnecessary" instanceof checks in TypeScript to make the code behave well when being used from JavaScript, which doesn't enforce type checking. I know we do this in the web sdk, but not so sure about in the node sdk.

Comment (tom-andersen) [reply dev/src/reference.ts:2865]: This actually has more to do with inheritance. Notice that we are dealing with `AggregateField<T>` and `firestore.AggregateField<T>`. Any other class that implements the supertype will not be considered equal. When more aggregate types are introduced in the future, there will also be additional comparison of internals that requires we first do `instanceof` to safely access properties.

File: dev/src/reference.ts
Comment (tom-andersen) [reply dev/src/reference.ts:3022]: Copy pasta! Removed.

File: dev/system-test/firestore.ts
Comment (tom-andersen) [reply dev/system-test/firestore.ts:2216]: Will do.

File: dev/src/reference.ts
Comment (tom-andersen) [reply dev/src/reference.ts]: Done

File: dev/src/reference.ts
Comment (ehsannas) [reply dev/src/reference.ts:2865]: Makes sense. Thanks

**Review by ehsannas — APPROVED** [review]
LGTM. I've added the "do not merge" label. Since the PR exposes the API publicly we should wait until allowed to merge the PR.

**Review by dconeybe — APPROVED** [review]

**Pull request merged**
